src/recordlinker/linkage/models.py

```diff
@@ -15,35 +29,125 @@ class Person(Base):
+    external_person_id: orm.Mapped[str] = orm.mapped_column(sqltypes.String(255), nullable=True)
+    external_person_source: orm.Mapped[str] = orm.mapped_column(sqltypes.String(100), nullable=True)
+    blocking_values: orm.Mapped[list["BlockingValue"]] = orm.relationship(back_populates="patient")
+
+
+class BlockingKey(enum.Enum):
+    """
+    Enum for the different types of blocking keys that can be used for patient
+    matching. This is the universe of all possible blocking keys that a user can 
+    choose from when configuring their algorithm.  When data is loaded into the
+    MPI, all possible BlockingValues will be created for the defined BlockingKeys.
+    However, only a subset will be used in matching, based on the configuration of
+    the algorithm.  By defining them all upfront, we give the user flexibility in
+    adjusting their algorithm configuration without having to reload the data.
+
+    **HERE BE DRAGONS**: IN A PRODUCTION SYSTEM, THESE ENUMS SHOULD NOT BE CHANGED!!!
+    """
+    BIRTHDATE = 1, "Date of Birth"
+    MRN = 2, "Last 4 chars of MRN"
+    SEX = 3, "Gender"
+    ZIP = 4, "Zip Code"
+    FIRST_NAME = 5, "First 4 chars of First Name"
+    LAST_NAME = 6, "First 4 chars of Last Name"
+
+    def __init__(self, id: int, description: str):
+        self.id = id
+        self.description = description
+
+    def to_value(self, data: dict) -> list[str]:
+        """
+        Given a data dictionary of Patient PII data, return a generator of all
+        possible values for this Key.  Many Keys will only have 1 possible value,
+        but some (like first name) could have multiple values.
+        """
+        if self == BlockingKey.BIRTHDATE:
+            return self._extract_birthdate(data)
+        if self == BlockingKey.MRN:
+            return self._extract_mrn_last_four(data)
+        if self == BlockingKey.SEX:
+            return self._extract_sex(data)
+        if self == BlockingKey.ZIP:
+            return self._extract_zipcode(data)
+        if self == BlockingKey.FIRST_NAME:
+            return self._extract_first_name_first_four(data)
+        if self == BlockingKey.LAST_NAME:
+            return self._extract_last_name_first_four(data)
+        return []
+
+    def _extract_birthdate(self, data: dict) -> list[str]:
+        if "birthdate" in data:
+            val = data["birthdate"]
+            if not isinstance(val, (datetime.date, datetime.datetime)):
+                # if not an instance of date or datetime, try to parse it
+                val = dateutil.parser.parse(str(val))
+            return [val.strftime("%Y-%m-%d")]
+        return []
+
+    def _extract_mrn_last_four(self, data: dict) -> list[str]:
+        if "mrn" in data:
+            return [data["mrn"].strip()[-4:]]
+        return []
+
+    def _extract_sex(self, data: dict) -> list[str]:
+        if "sex" in data:
+            val = str(data["sex"]).lower().strip()
```

Comment:
some of this, like lower() and strip(), will be moved to the parsing section of the new PIIRecord serializer model I'm working on.  I'm trying to separate the concerns of parsing a dict/json string into a structure from extracting some value.

Since the current code just uses a generic python dict, we kind of have to parse and extract in every use case (we aren't obligated, but thats the safest thing to do).  Converting the "data: dict" to a "data: PIIRecord" will make this easier on the eyes.